es5-shim.js

```diff
@@ -329,6 +351,94 @@ var strSplit = call.bind(StringPrototype.split);
 var strIndexOf = call.bind(StringPrototype.indexOf);
 var push = call.bind(array_push);
 
+// ES-5 15.3.4.4
+// http://es5.github.io/#x15.3.4.4
+var hasToStringTagBasicBug = toStr() !== '[object Undefined]' || toStr(null) !== '[object Null]';
+var hasToStringTagLegacyArguments = toStr(function () {
+  return arguments;
+}) !== '[object Arguments]';
+var hasToStringTagArrayBug = toStr([]) !== '[object Array]';
+var hasToStringTagRegExpBug = toStr(/test/) !== '[object RegExp]';
+var hasToStringTagInconsistency = hasToStringTagBasicBug ||
+    hasToStringTagLegacyArguments ||
+    hasToStringTagArrayBug ||
+    hasToStringTagRegExpBug;
+
+// Others to check:
+// Safari 8 returns 'object' for typed array constructors-
+// PhantomJS 1.9 which returns 'function' for `NodeList` instances.
+
+if (hasToStringTagLegacyArguments) {
+  var isDuckTypeArguments = function (value) {
+    return value != null &&
+        typeof value === 'object' &&
+        owns(value, 'length') &&
+        typeof value.length === 'number' &&
+        value.length >= 0 &&
+        owns(value, 'callee') &&
+        isCallable(value.callee) &&
+        !owns(value, 'arguments');
+
+  };
+}
+
+if (hasToStringTagArrayBug) {
+  var isDuckTypeArray = function (value) {
+    return value != null &&
+        typeof value === 'object' &&
+        owns(value, 'length') &&
+        typeof value.length === 'number' &&
+        value.length >= 0 &&
+        !owns(value, 'callee') &&
+        !owns(value, 'arguments');
+  };
+}
+
+if (hasToStringTagRegExpBug) {
+  var isTryCatchRegExp = function (value) {
+    if (value != null) {
+      var type = typeof value;
+      if (type === 'object' || type === 'function') {
+        try {
+          regexExec.call(value);
+          return true;
+        } catch (e) {
+          return false;
+        }
+      }
+    }
+  };
+}
+
+if (hasToStringTagInconsistency) {
+  defineProperties(FunctionPrototype, {
+    call: function () {
```

Comment:
Yeah, a similar thing will need to be done with `apply`. Just concentrating on `call` right now, exploring the possibilities, proving that it can be done.